Reset-Office365Tenant/Office365.psm1

```diff
@@ -0,0 +1,377 @@
+        Write-Host -ForegroundColor Red "ERROR: Get-ManagementRoleAssignment failed with $_"
+        return $null
+    }
+}
+
+############################################################
+# Enable OrganizationCustomization for the tenant if not hydrated
+############################################################
+function Enable-TenantCustomization()
+{
+    try
+    {
+        DisplayInProgressOperation "Enable OrganizationCustomization"
+        Enable-OrganizationCustomization -ErrorAction silentlycontinue
+        DisplayCompletedOperation "OK" Green
+        return $true
+    }
+    catch
+    {
+        DisplayCompletedOperation "FAIL" Red
+        Write-Host -ForegroundColor Red "ERROR: Enable-OrganizationCustomization failed with $_"
+        return $false
+    }
+}
+
+############################################################
+# Add Impersonation Rights to a user
+############################################################
+function Add-ImpersonationRights($username)
+{
+    try
+    {
+        DisplayInProgressOperation "Set Impersonation rights"
+        New-ManagementRoleAssignment -Role "ApplicationImpersonation" -User  $username
+        DisplayCompletedOperation "OK" Green
+        return $true
+    }
+    catch
+    {
+        DisplayCompletedOperation "FAIL" Red
+        Write-Host -ForegroundColor Red "ERROR: Set-ManagementRoleAssignment failed with $_"
+        return $false
+    }
+}
+
+################################################################################
+# Confirms if user has Impersonation Rights
+################################################################################
+function Confirm-ImpersonationRights($username)
+{
+    # Test if tenant is dehydrated
+    $isDehydrated = Test-IsTenantDehydrated
+
+    # Fail if we cannot get this information
+    if ($null -eq $isDehydrated)
+    {
+        return $false
+    }
+    
+    # Enable-OrganizationCustomization if tenant is dehydrated
+    if ($true -eq $isDehydrated)
+    {
+        if ($false -eq $(Enable-TenantCustomization))
+        {
+            return $false
+        }
+    }
+
+    # Test if user has Impersonation rights
+    $hasRights = Test-HasImpersonationRights -username $username
+    if ($false -eq $hasRights)
+    {
+        # Add Impersonation rights
+        return Add-ImpersonationRights -username $username
+    }
+    return $true
+}
+
+################################################################################
+# Reset Office365 Tenant
+################################################################################
+function Reset-Office365Tenant($credentials, $defaultPassword, $maxUser)
+{
+    if ($null -eq $credentials)
+    {
+        Write-Host "Ask for credentials"
+        $credentials = Get-Credentials -Message "Please insert your Office365 Admin Credentials" 
+    }
+    else
+    {
+        Write-Host "Credentials supplied"
+    }
+
+    # Clean all PSSessions
+    Clear-PSSession
```

Comment:
If this function fails it returns false, but you never handle what happens after that. It doesn't appear to be critical to the rest of the flow that this completes, but it would be nice to clarify what should ultimately happen when this fails.

Reply:
If can't log into Office 365, bail with an error message :)